presto-main/src/main/java/com/facebook/presto/util/GcStatusMonitor.java

```diff
@@ -0,0 +1,202 @@
+import com.facebook.presto.execution.TaskStatus;
+import com.facebook.presto.memory.QueryContext;
+import com.facebook.presto.operator.TaskStats;
+import com.facebook.presto.spi.QueryId;
+import com.facebook.presto.spi.memory.MemoryPoolInfo;
+import io.airlift.log.Logger;
+import io.airlift.stats.GarbageCollectionNotificationInfo;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.management.JMException;
+import javax.management.Notification;
+import javax.management.NotificationListener;
+import javax.management.ObjectName;
+import javax.management.openmbean.CompositeData;
+
+import java.lang.management.GarbageCollectorMXBean;
+import java.lang.management.ManagementFactory;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+
+public class GcStatusMonitor
+{
+    private final Logger log = Logger.get(GcStatusMonitor.class);
+    private final NotificationListener notificationListener = (notification, ignored) -> onNotification(notification);
+    private final SqlTaskManager sqlTaskManager;
+
+    @Inject
+    public GcStatusMonitor(
+            SqlTaskManager taskManager)
+    {
+        this.sqlTaskManager = taskManager;
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().addNotificationListener(
+                        objectName,
+                        notificationListener,
+                        null,
+                        null);
+            }
+            catch (JMException e) {
+                throw new RuntimeException("Unable to add GC listener", e);
+            }
+        }
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().removeNotificationListener(objectName, notificationListener);
+            }
+            catch (JMException ignored) {
+            }
+        }
+    }
+
+    private synchronized void onNotification(Notification notification)
+    {
+        if ("com.sun.management.gc.notification".equals(notification.getType())) {
+            GarbageCollectionNotificationInfo info = new GarbageCollectionNotificationInfo((CompositeData) notification.getUserData());
+            if (info.isMajorGc()) {
+                onMajorGc();
+            }
+        }
+    }
+
+    private void onMajorGc()
+    {
+        try {
+            onMajorGcLogic();
+        }
+        catch (Throwable throwable) {
+            log.error(throwable);
+        }
+    }
+
+    private void onMajorGcLogic()
```

Comment:
onMajorGCLogging?

I suspect this name will trigger someone to ask for a change.